API: Rename `numpy/core` to `numpy/_core` [NEP 52]

mtsokol opened the pull request:
Hi @rgommers,

This PR introduces a breaking change of renaming `numpy/core` module to `numpy/_core`. The first commit renames the module, where following ones adjust the codebase to this change. 

Some stages will fail for now (e.g. building docs, as matplotlib needs to have some changes introduced), but I checked all 2500 occurrences of `core` in the code 😄 and this PR should cover all of the relevant ones.

mtsokol committed:
MAINT: Rename numpy.core module to numpy._core

mtsokol committed:
MAINT: Update numpy.core paths to numpy._core

MAINT: Update git submodules, fix test_format

rgommers commented on numpy/tests/test_public_api.py:
```diff
@@ -164,21 +165,21 @@ def test_NPY_NO_EXPORT():
-    "core.records",
-    "core.shape_base",
-    "core.umath",
+    "_core",
```
All of these can be deleted instead of renamed I believe - they should simply not show up anymore.

rgommers commented on tools/refguide_check.py:
```diff
@@ -466,7 +466,7 @@ def resolve(name, is_label=False):
     token = '<RST-VALIDATE-SYNTAX-CHECK>'
 
-    docutils.core.publish_doctree(
+    docutils._core.publish_doctree(
```
Incorrect change?

mtsokol replied:
Ah, yes - I left `docutils.core` as is but this one went through.

mtsokol replied:
Fixed.

rgommers commented on numpy/lib/tests/test_format.py:
```diff
@@ -536,17 +536,14 @@ def test_python2_python3_interoperability():
         data = np.load(path)
     assert_array_equal(data, np.ones(2))
 
-def test_pickle_python2_python3():
```
What's happening with these code removals?

mtsokol replied:
I had to update array files in the `numpy/lib/tests/data/` folder, because otherwise these tests are failing with "`numpy.core.multiarray` doesn't exist".  
When you `cat` these `.npy` and `.npz` files, in their second lines you will see `numpy._core.multiarray`. Numpy's module path is there so I had to recreate these files with `np.save`.  

For `Python 2` files (`py2-objarr.npy` and `py2-objarr.npz`) I would need to be able to run the codebase with Python 2 to also get updated versions of these zips, but as I understand Python 2 is no longer supported. 

Therfore I decided to drop test cases where we test against Python 2. Is it correct? Or should we get them anyway?

P.S. I tried to manually change `core` to `_core` in vim for Python 2 `npy` and `npz` files but then, as expected, I get a "checksum error" on loading them.

rgommers replied:
I already had to do such `.npy`/`.npz` file updates recently: https://github.com/PyWavelets/pywt/pull/679. It seems to have been a problem in numpy 1.25.2 already.

This change is larger, and hence more "saved in python 2.X" files stop working. They'll probably start working again if you put back `numpy/core/__init__.py` with a deprecation warning. But either way, eventually files with object arrays are going to stop working indeed. Technically pickled data isn't portable between Python versions at all IIRC. I don't think it affects too many users of `.npy` files, only those with object arrays.

@rkern do you have any objections to dropping these tests?

rkern replied:
Yes, pickled data is portable between Python versions, as long as both versions support the same protocol level. [It's one of the defining differences between it and `marshal`](https://docs.python.org/3/library/pickle.html#comparison-with-marshal).

I do object to dropping these tests. I would like these old files to continue to work, as they exist. If you want updated versions of some of these files to test with the `numpy._core` names, then please make them *in addition* to the existing files. How the unpickler resolves names can be customized by overriding [`Unpickler.find_class()`](https://docs.python.org/3/library/pickle.html#pickle.Unpickler.find_class), and we should do so where we do use unpickling in the NPY/NPZ formats in order to follow this renaming.

Whether we expose this customized `Unpickler` for others to use for their own unpickling to handle this refactoring, I am neutral on. But for the NPY/NPZ formats, we abstracted away many of these details from the user, and I think we should take this relatively small step to ensure that our internal refactorings do not invalidate their data.

rgommers replied:
Ah, that's a good idea, thanks. +1 for an unpickler.

mtsokol replied:
Sure! I reintroduced python2 pickle tests and managed to get them passing by extending `core` shim that is used when `npy`/`npz` file was pickled with `numpy.core` module.

rgommers reviewed (commented):
Thanks @mtsokol! Wow, this turned out to be a large PR. A few initial comments. Let's see what is needed in downstream PRs to stop using `numpy.core` completely. I believe we have to put `numpy/core/__init__.py` back with a deprecation warning emitted when accessed, and a `__getattr__` to forward imports to the relevant `numpy._core` items. Otherwise this will probably be the most disruptive change in the release.

ksunden commented:
> (e.g. building docs, as matplotlib needs to have some changes introduced)

It's not actually obvious to me why this is failing, we don't import `np.core` anywhere.

The failure is on import of a C-extension module, which does depend on the Numpy C API via:

```c
#include <numpy/ndarrayobject.h>

and

#include "numpy/arrayobject.h"
```

None of our C source files refer to "core" directly, either.

So quite possibly it is "just" a matter of recompiling... but that is itself a potential problem if this affects the ability to have binaries compatible with different numpy versions. (keeping existing binaries forward-compatible would be ideal, if possible)

I do suspect that PyBind11 will have problems that need to be addressed (which may be prerequisites for even testing what, if anything, needs to be done for mpl specifically). A quick scan of the source there shows that they do refer to `numpy.core` in at least a handful of places. The particular c-extension that is failing is _not_ a pybind11 module, but we do have some which _are_ (and are slowly working to increase usage of pybind11).

ksunden commented:
Pandas also appears to be affected (there are some tests which refer to `np.core` directly, but nothing in production code), but also from the doc build, a pandas Cython extension module is broken at `cnp.import_array()` (`cnp` just being the `cimport numpy as cnp`)

rgommers commented:
> So quite possibly it is "just" a matter of recompiling... but that is itself a potential problem if this affects the ability to have binaries compatible with different numpy versions. (keeping existing binaries forward-compatible would be ideal, if possible)

I think that that should be possible _for this PR_, and that's probably the number one prio to sort out here. However, I do want to point out that the plan is for NumPy 2.0 to break ABI compatibility - so keeping existing binaries forward-compatible during this release cycle is highly unlikely to happen. See the guidance in gh-24300.

ksunden commented:
Hmmm... well that raises some infrastructure/CI questions during the transition... if matplotlib (and pandas) is required for numpy docs, but matplotlib needs to be rebuilt due to ABI changes. Perhaps as backwards incompatible changes roll in, you need to install from source to avoid a chicken/egg problem. (Though if/when changes are needed to pybind11/cython, etc, we may not escape the problem entirely)

Perhaps the way to do it for mpl is to pin to `<2.0` on the release branches, but leave unpinned on main, so that makes it easier for devs/numpy CI to install with 2.0-compiled versions without having to edit the dependency metadata.

rgommers commented:
> but matplotlib needs to be rebuilt due to ABI changes. Perhaps as backwards incompatible changes roll in,

We won't be breaking things all the time - C API/ABI changes are being prepared behind a switch. 

> Perhaps the way to do it for mpl is to pin to `<2.0` on the release branches, but leave unpinned on main, so that makes it easier for devs/numpy CI to install with 2.0-compiled versions without having to edit the dependency metadata.

Yes, that sounds exactly right. I just checked the latest Matplotlib release (3.7.2), and it uses `numpy>=1.20` in its `setup.py`. That is going to break - you should start adding `<2.0`. I've tried to communicate this in as many places as possible (numpy mailing list, 2.0 tracking issue, https://numpy.org/devdocs/dev/depending_on_numpy.html, other issues and meetings), but it feels like no one has seen it. So it's time to get more proactive and find non-numpy venues (blogs/podcast/etc.) to flag this (as @mattip suggested recently).

mtsokol committed:
MAINT: Expand numpy.core shims

mtsokol commented:
> > (e.g. building docs, as matplotlib needs to have some changes introduced)
> 
> It's not actually obvious to me why this is failing, we don't import `np.core` (directly) anywhere. ...

@ksunden I managed to solve the issue - it was due to the fact that matplotlib used in docs is compiled against numpy with `numpy.core` and [imports from `core` as a part of numpy initialization](https://github.com/numpy/numpy/blob/1ed0bfd3a42bf0c8a0060684645ba97941778006/numpy/core/code_generators/generate_numpy_api.py#L47). I added `numpy.core` shims as a proxy to `numpy._core` for a full backward compatibility.

rgommers commented:
This LGTM:
```python
>>> np.core.multiarray.concatenate
<ipython-input-4-423c639bb0c0>:1: DeprecationWarning: `numpy.core` has been made officially private and renamed to `numpy._core`. Accessing `_core` directly is discouraged, you should use a public module instead that exports the attribute in question. If you still would like to access an attribute from it, please use `np._core.multiarray`.
  np.core.multiarray.concatenate
<function concatenate at 0x7f5ecca17630>

>>> np.core._multiarray_umath.arange
<ipython-input-5-10613df9b3ac>:1: DeprecationWarning: `numpy.core` has been made officially private and renamed to `numpy._core`. Accessing `_core` directly is discouraged, you should use a public module instead that exports the attribute in question. If you still would like to access an attribute from it, please use `np._core._multiarray_umath`.
  np.core._multiarray_umath.arange
<built-in function arange>
```

Everything I tried still works, with a deprecation message appearing. The thing I'm wondering about too is if the reverse is going to keep working. If the one problem here was:
```C
PyObject *numpy =  PyImport_ImportModule("numpy.core._multiarray_umath");
```
and now we rebuild MPL against a new numpy version where that extension is in `_core`, can the C API shim or `numpy2_compat` package from https://numpy.org/neps/nep-0053-c-abi-evolution.html fix that? If not, we may have special-case that to ensure that that particular import works without a deprecation message - maybe continue baking the `core` names into the extension modules (that doesn't matter much in the end after all, as long as we get the deprecation messages the way we expect them to appear).

mtsokol commented:
> Everything I tried still works, with a deprecation message appearing. The thing I'm wondering about too is if the reverse is going to keep working. If the one problem here was:
> 
> ```c
> PyObject *numpy =  PyImport_ImportModule("numpy.core._multiarray_umath");
> ```
> 
> and now we rebuild MPL against a new numpy version where that extension is in `_core`, can the C API shim or `numpy2_compat` package from https://numpy.org/neps/nep-0053-c-abi-evolution.html fix that? 

@rgommers I don't think there's anything that needs to be fixed. If a C file imports `numpy.core._multiarray_umath` from a NumPy with `_core`, it will work as we have `core` shim in Python, I checked it myself (Python's deprecation warning is not displayed during C execution).

rgommers commented:
> Did I get your question right?

Not quite, I meant the reverse. If we build Matplotlib against NumPy 2.0, and hence the C code (after this PR) uses `numpy._core._multiarray_umath`, that won't work against NumPy 1.25.0 because there's no `numpy._core` there. So my question is how we make that work.

Another thought: it could be done through try-except logic in the C code: if `numpy._core` doesn't exist, then fall back to `numpy.core`. Or import `numpy` first, check the version logic, and then decide if we need `core` or `_core`.

mtsokol commented:
> > Did I get your question right?
> 
> Not quite, I meant the reverse. If we build Matplotlib against NumPy 2.0, and hence the C code (after this PR) uses `numpy._core._multiarray_umath`, that won't work against NumPy 1.25.0 because there's no `numpy._core` there. So my question is how we make that work.
> 
> Another thought: it could be done through try-except logic in the C code: if `numpy._core` doesn't exist, then fall back to `numpy.core`. Or import `numpy` first, check the version logic, and then decide if we need `core` or `_core`.

Ah, now it's clear: Then it's either by trying to import or checking numpy version directly in `import_array()`.

seberg commented:
Changing from where we import based on NumPy is not a problem.  It would be good if the old place keeps working modulo a possible deprecation warning.  That just makes our life easier right now where we are not actually forcing recompilation (but just hoping that the ABI changes are small enough that downstream is fine).

mtsokol committed:
MAINT: Check numpy version in C-API

mtsokol committed:
MAINT:  Adjustments after numpy._core rename

Match x86-simd-sort with the main branch

Add remaining shims and a test

Adjust docstrings and imports for isalpha

Match svml version with the main branch

Apply review changes

Rework np.core stubs to generate warnings more often

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

ngoldbaum commented:
the pybind11 PR hot merged, does that unblock downstream users who depend on pybind11? It’s been a while since I’ve used pybind11 - do we need to worry about downstream projects vendoring it?

It looks like some merge conflicts have accumulated. This PR is big enough that conflicts are inevitable. @mtsokol feel free to resolve the conflicts but it might make sense to wait and do it all at once right before we merge this.

mtsokol commented:
> the pybind11 PR hot merged, does that unblock downstream users who depend on pybind11? It’s been a while since I’ve used pybind11 - do we need to worry about downstream projects vendoring it?

I think we're good to go. 
The fix in pybind11 was meant to avoid a warning when accessing `np.core...` only. It can be easily avoided on the user side (including users vendoring pybind11) by suppressing it (I already did it in scipy to avoid nightly build errors: https://github.com/scipy/scipy/pull/19275). 

> It looks like some merge conflicts have accumulated. This PR is big enough that conflicts are inevitable. @mtsokol feel free to resolve the conflicts but it might make sense to wait and do it all at once right before we merge this.

Sure! I rebased the PR.

ngoldbaum commented on numpy/_core/include/numpy/experimental_dtype_api.h:
```diff
@@ -353,7 +354,12 @@ import_experimental_dtype_api(int version)
+        PyErr_Clear();
+        multiarray = PyImport_ImportModule("numpy.core._multiarray_umath");
+    }
+
```
I don't care much but I think it's OK to only import from `_core` in this header, it's experimental after all.

mtsokol replied:
Make sense - I removed fallback import.

ngoldbaum commented on numpy/_core/tests/test_mem_policy.py:
```diff
@@ -290,30 +290,32 @@ class MyArr(np.ndarray):
 
 
 async def concurrent_context2(get_module, orig_policy_name, event):
+    from numpy._core.multiarray import get_handler_name
```
If you want to make this simpler I don't think there's any problem with importing `numpy._core.multiarray` at top-level in this file. 

Please also keep how all of these tests are structured consistent, if you want to remove all the explicit `numpy._core.multiarray` do that in every test, not just some.

mtsokol replied:
Sure! I moved `numpy._core.multiarray` to a top-level import in files where it was imported more than once in unit tests.

ngoldbaum commented on numpy/_core/tests/test_multiarray.py:
```diff
@@ -4321,44 +4321,89 @@ def _loads(self, obj):
+            b"ndarray\nq\x02K\x00\x85U\x01b\x87Rq\x03(K\x04\x85cnumpy\ndt"
+            b"ype\nq\x04U\x02f4K\x00K\x01\x87Rq\x05(U\x01<NNJ\xff\xff\xff"
+            b"\xffJ\xff\xff\xff\xfftb\x89U\x10\x00\x00\x80?\x00\x00\x00@"
+            b"\x00\x00@@\x00\x00\x80@tb."
```
I think it's simpler to leave all of this on one line. Also all the other tests where this happened in this file.

mtsokol replied:
Ok, I reverted these changes and added `# noqa` at the end to avoid linter errors.

ngoldbaum commented on numpy/_core/tests/test_regression.py:
```diff
@@ -458,23 +458,24 @@ def test_pickle_py2_bytes_encoding(self):
             # (original, py2_pickle)
             (np.str_('\u6f2c'),
-             b"cnumpy.core.multiarray\nscalar\np0\n(cnumpy\ndtype\np1\n"
+             b"cnumpy._core.multiarray\nscalar\np0\n(cnumpy\ndtype\np1\n"
```
This change makes me realize we might also need to work on NumPy's pickle code so old pickle files don't generate deprecation warnings that downstream users can do nothing about without recreating the pickle file.

ngoldbaum replied:
IMO all of these regression tests should pass untouched with the old `numpy.core` paths. Hopefully the pickle API allows NumPy to be smart enough to replace that path inline while loading the pickle.

mtsokol replied:
I remember testing it - if a pickle contains `numpy.core.*` due to the fact that it was pickled with 1.x, then unpickling with 2.x emits a warning. 

I will revert changes here and add filter warning for now.I propose I will work on warning-less unpickling for these python 2 formats in a follow-up PR, is it OK?

ngoldbaum replied:
> I will revert changes here and add filter warning for now.I propose I will work on warning-less unpickling for these python 2 formats in a follow-up PR, is it OK?

SGTM, I just opened https://github.com/numpy/numpy/issues/24844 to track this. 

It would be nice if we could check that it is actually feasible to handle this inside NumPy before merging this PR, but no need to block merging this until the fix is implemented. It may be the case that the imports happen inside of the pickle internals and there is no way to NumPy to intercept the import.

ngoldbaum commented on numpy/_core/umath.py:
```diff
@@ -0,0 +1,33 @@
+Create the numpy._core.umath namespace for backward compatibility. In v1.16
+the multiarray and umath c-extension modules were merged into a single
+_multiarray_umath extension module. So we replicate the old namespace
+by importing from the extension module.
```
I wonder if we need this module in `numpy._core` - is numpy internally still using these imports? I guess it would make the renaming a bit more complicated if downstream is using things from this namespace on purpose, but also almost all of this stuff is available in the main namespace.

mtsokol replied:
When I search for `_core.umath` in the codebase it's used/imported in a number of places (especially in `code_generators`). I see that there are also downstream usages but this PR provides shim for `numpy.core.umath` so downstream will still work (with a deprecation warning).
My guess would be to leave this module.

ngoldbaum replied:
SGTM

ngoldbaum commented on numpy/core/__init__.py:
Please include a top-level docstring that indicates that this file exists purely for backward compatibility and should be removed in the future.

mtsokol replied:
Added!

ngoldbaum commented on numpy/core/_utils.py:
```diff
@@ -0,0 +1,14 @@
+        "`numpy._core`. Accessing `_core` directly is discouraged, you "
+        "should use a public module instead that exports the attribute "
+        "in question. If you still would like to access an attribute "
+        f"from it, please use `np._core.{attr}`.",
```
Also add a line warning them that things in `np._core` can change without warning at any time.

mtsokol replied:
Good point - I added a line that says it can change without a warning.

ngoldbaum commented on numpy/polynomial/hermite.py:
```diff
@@ -77,7 +77,7 @@
 import numpy as np
 import numpy.linalg as la
-from numpy.core.multiarray import normalize_axis_index
+from numpy._core.multiarray import normalize_axis_index
```
This is publicly visible from `numpy.lib.array_utils`. I think it might be worthwhile to change internal imports from `_core` into imports from the public API. Not a blocker though. That said, doing this exercise might also be a good way to test your `ruff` PR.

ngoldbaum replied:
`ruff` would need to be able to suggest changing `numpy.core.absolute` to `numpy.absolute` but `numpy.core.multiarray.normalize_axis_index` to `numpy.lib.array_utils.normalize_axis_index`. There's probably lots of usages like this downstream so it would be nice if we could auto-fix them.

mtsokol replied:
I changed these imports to `numpy.lib.array_utils.*`.

Yes, I will include it in ruff rule.

ngoldbaum commented on numpy/tests/test_public_api.py:
```diff
@@ -289,17 +280,17 @@ def is_unexpected(name):
+    "numpy._core.code_generators.verify_c_api_version",
+    "numpy._core.cversions",
+    "numpy._core.generate_numpy_api",
 ]
```
Are these actually needed anymore now that `_core` is private? I think you might just be able to delete all of this.

mtsokol replied:
Right, underscore in the name already makes these modules as skippable.

ngoldbaum commented on tools/refguide_check.py:
```diff
@@ -60,7 +60,7 @@
 
 PUBLIC_SUBMODULES = [
-    'core',
+    '_core',
```
Not public anymore. That said there's lots of public types with docstrings defined inside of `_core` so maybe it's correct to leave this here? Someone with more experience with this file should weigh in.

mtsokol replied:
Right, this one should be removed.

ngoldbaum reviewed (commented):
I don't think there's much we can do about it, but this will make git archeology on anything in `numpy.core` marginally more annoying with the renames, particularly for the C code. Just want to note that.

It looks like you broke up some long lines, presumably because of the linter? While most of these are readability improvements, I think some of these long lines are preferable as-is and just need a `# noqa` comment at the end of the line so `pycodestyle` ignores it. See inline comments.

I understand the logic for keeping a stub `np.core` module that exists purely to forward imports and raise a deprecation warning. Can you double-check that all functionality that's been moved to `numpy._core` is still importable with a warning (even from a submodule)? I think it's fine but worry slightly that there's a missing submodule that doesn't forward imports. It might also warrant a test if there are common semi-public things that people like to import and we don't want to break for a few releases.

mtsokol commented on doc/Py3K.rst:
Sure! I will remove it in a follow-up PR.

mtsokol commented on doc/release/upcoming_changes/24634.change.rst:
```diff
@@ -0,0 +1,6 @@
+``numpy/core`` is now officially a private module and access to it
+in daily development is discouraged. To ensure backward compatibility,
+``numpy/core`` is still present as a shim that emits
+a ``DeprecationWarning`` on access.
```
Sure! I used this description.

mtsokol commented on numpy/_core/__init__.pyi:
```diff
@@ -0,0 +1,2 @@
+# NOTE: The `np._core` namespace is deliberately kept empty due to it
+# being private (despite the lack of leading underscore)
```
Fixed!

mtsokol commented on numpy/_core/_add_newdocs.py:
```diff
@@ -3690,9 +3698,16 @@
+        keepdims=False, 
+        initial=1, 
+        where=True
+    )
```
I did it as linter complained there - sure, I reverted it to a one-line.

mtsokol commented on numpy/_core/_dtype_ctypes.py:
Correct. In the first commit in this PR I just renamed all files, and github displayed it correctly. But as I added shims (files for forwarding content with a deprecation) in `numpy.core` github now picks it as a modification of a original file and addition of a new one (thus ~7000 lines added/removed).

mtsokol commented:
> It looks like you broke up some long lines, presumably because of the linter? While most of these are readability improvements, I think some of these long lines are preferable as-is and just need a `# noqa` comment at the end of the line so `pycodestyle` ignores it. See inline comments.

@ngoldbaum Yes, I split these raw byte strings to pass linting stage. I reverted them to the original shape with `# noqa` at the end.

> I understand the logic for keeping a stub `np.core` module that exists purely to forward imports and raise a deprecation warning. Can you double-check that all functionality that's been moved to `numpy._core` is still importable with a warning (even from a submodule)? I think it's fine but worry slightly that there's a missing submodule that doesn't forward imports. It might also warrant a test if there are common semi-public things that people like to import and we don't want to break for a few releases.

Sure! I provided shims for all "former-semi-public" submodules residing in the former `numpy.core`. By "former-semi-public" I mean these members, whose names don't start with `_`.  
Additionally, I created a test named `test_core_shims_coherence` in the `test_public_api.py` which verifies that every "former-semi-public" `_core` member (including "former-semi-public" members in `_core`'s submodules) can be imported from `numpy.core.*` shims. The test passes.

ngoldbaum reviewed (approved):
This looks good to me now! 

Thank you for applying the review comments in a separate commit, makes it much easier to review that way. 

This is an enormous diff - although mostly just renames - I'd like it if someone else could look over this before merging. I'll try to make sure this comes up at the triage review this week.

mtsokol commented:
Great! Let me add changelog file also in another commit.

ngoldbaum commented:
This may have gotten lost:

> It would be nice if we could check that it is actually feasible to handle this inside NumPy before merging this PR, but no need to block merging this until the fix is implemented. It may be the case that the imports happen inside of the pickle internals and there is no way to NumPy to intercept the import.

Before this is merged, can you check if it is actually possible for us to fix this inside NumPy? If the import happens inside the pickle internals then there might be nothing we can do about it and we would need to have a bigger conversation about how to handle this issue. There are *a lot* of pickle files with numpy data structures inside of them in circulation - apparently it's [popular to share machine learning model weights](https://huggingface.co/docs/hub/security-pickle#mitigation-strategies) as a pickle file :upside_down_face:

mtsokol commented:
@ngoldbaum I checked it and the warning is raised from the cpython side (somewhere from [here](https://github.com/python/cpython/blob/6387b5313c60c1403785b2245db33372476ac304/Modules/_pickle.c#L7803)). IMO there's no control over it in numpy - pickle contains package path and pickle uses it to import class, therefore we get a warning if `numpy.core` was used in pickling. We have two choices:

1. Keep warning when unpickling pickle from previous versions (undesired as I understand)
2. Remove warning from **only one submodule shim** - `numpy.core.multiarray` - I think this is acceptable and a relatively small price. I don't see other way to avoid the warning message.

WDYT?

ngoldbaum commented:
> WDYT?

I'd probably lean towards loading old pickle files generating deprecation warnings. In a few releases `np.core` will be removed and if there is no way for us to fix this issue when loading the pickle in-place then loading the file will *break* at that point, unless we leave behind a semi-permanent compatibility shim. Worse there will be other usages of `np.core` that aren't invoked by pickle that should see deprecation warnings but won't if we turn it off for a submodule to accommodate pickle.

I don't think pickle should ever be used to share data outside of a controlled environment where all hosts opening the pickle file are fully controlled by a single entity. I don't think NumPy should bend over backwards to keep pickle files working between major package releases.

All that said, I will make sure this gets discussed on Wednesday.

mtsokol commented:
Sure! Then I leave for now regression tests as they are - with deprecation warning filter.

rkern commented:
One thing that I think we _should_ bend over backwards to do is to provide a customized `Unpickler` that can load the old files even after we carry through the deprecation and stop just warning and begin to just raise exceptions. [We will need one to support old NPY files](https://github.com/numpy/numpy/pull/24634/files#r1315137174) that happened to use pickling because of object arrays (because we hid the pickling behind an abstraction, I feel we have more duties to continue to handle old data than arbitrary pickles made directly with `pickle.dump()`).

ngoldbaum commented:
That's a good point! For `np.load` maybe it'll be easier to detect and handle this inside NumPy so users don't need to worry about it.

mtsokol commented:
@ngoldbaum @rkern Thank you for the feedback! In the latest commit [70b0127](https://github.com/numpy/numpy/pull/24634/commits/70b0127dd08045886dd7cdd85b54d43134186116) I introduced new `NumpyUnpickler` that is available in `numpy.lib.format` module. It allows to load pickles created with numpy 1.x without a warning and to **load them once `numpy.core` is gone** (it's a thin wrapper for `pickle.Unpickler` that rewires `numpy.core` name).

With `NumpyPickler`, tests in question from `test_regression.py` and `test_format.py` pass without a warning and with original contents.

If you have better name/location for this class I'm happy to move it!

rkern commented on numpy/lib/format.pyi:
```diff
@@ -20,3 +20,6 @@ def read_array_header_2_0(fp): ...
 def read_array(fp, allow_pickle=..., pickle_kwargs=...): ...
 def open_memmap(filename, mode=..., dtype=..., shape=..., fortran_order=..., version=...): ...
+
+class NumpyUnpickler():
```
I'm not too familiar with these stub files, but I'd expect this to just subclass from `pickle.Unpickler`.

mtsokol replied:
That's right - I updated it.

ngoldbaum commented:
That sounds great!

Did you look at the issue with `np.save` and `np.load` with `allow_pickle=True`? Should we use this new unpickler there to avoid spurious warnings loading old .npy file

The migration guide should also be updated to demonstrate how to use it. I would show an example of how to load an old pickle file and how to create a new pickle file that doesn't contain the problematic `np.core` paths after loading the old data.

Also this is a new public API addition so it needs to appear in the API listing and it should be mentioned in the I/O docs, [just a sentence or two here, I think](https://numpy.org/doc/stable/user/how-to-io.html#save-restore-using-a-pickle-file), and the release note should mention it as well.

mtsokol commented:
> Did you look at the issue with `np.save` and `np.load` with `allow_pickle=True`? Should we use this new unpickler there to avoid spurious warnings loading old .npy file

@ngoldbaum when `allow_pickle=False` option is used, then any attempt to operate on pickles raises an exception on purpose. When `True` is provided it now uses `NumpyUnpickler` in `np.load`, so the warning doesn't occur.  
 
> The migration guide should also be updated to demonstrate how to use it. I would show an example of how to load an old pickle file and how to create a new pickle file that doesn't contain the problematic `np.core` paths after loading the old data.
> 
> Also this is a new public API addition so it needs to appear in the API listing and it should be mentioned in the I/O docs, [just a sentence or two here, I think](https://numpy.org/doc/stable/user/how-to-io.html#save-restore-using-a-pickle-file), and the release note should mention it as well.

Sure! I included all these changes in the latest commit.

ngoldbaum commented on doc/source/user/how-to-io.rst:
```diff
@@ -319,6 +319,11 @@ Use :func:`numpy.save` and :func:`numpy.load`.  Set ``allow_pickle=False``,
+:func:`numpy.load` also supports unpickling files created with NumPy 1.x.
+If you try to unpickle a 1.x pickled array with an object dtype directly, you
+will get a deprecation warning and, once ``numpy.core`` is removed, an exception.
+Use ``numpy.lib.format.NumpyUnpickler`` for unpickling these files.
```
```suggestion
Use :class:`numpy.lib.format.NumpyUnpickler` for unpickling these files.
```

This should make the reference appear as a cross-reference to the API docs

ngoldbaum commented on numpy/lib/format.py:
```diff
@@ -1011,3 +1011,15 @@ def isfileobj(f):
+class NumpyUnpickler(pickle.Unpickler):
+    """
+    A thin wrapper for :py:class:`pickle.Unpickler` that
+    allows to load legacy array pickles with ``dtype=object``.
```
```suggestion
    allows to load legacy array pickles.
```

It's not just `dtype=object` that's problematic for generic pickle files.

ngoldbaum commented on doc/source/numpy_2_0_migration_guide.rst:
```diff
@@ -154,3 +154,26 @@ ptp                     Use ``np.ptp(arr, ...)`` instead.
+  >>> new_pickle_file = io.BytesIO()
+  >>>
+  >>> arr = NumpyUnpickler(old_pickle_file).load()  # equivalent to pickle.load(file)
+  >>> # for pickling, a default Pickler can be used:
```
```suggestion
  >>> # save a new pickle file that doesn't need to be loaded with NumpyUnpickler:
```

ngoldbaum reviewed (commented):
Thanks! Just a couple minor tweaks to some of the language in the last commit.

mattip reviewed (commented):
I looked through this and didn't see things that jumped out at me. The tricky parts are around the public facing API: adding the warning when importing from `core` and the C-API changes in `_import_array` and friends. Is that code tested? I am worried that this change will be very noisy for downstream libraries. Do we have a handle on how many changes scipy/numexpr/numba would have to do to get clean tests?

mtsokol commented:
> I looked through this and didn't see things that jumped out at me. The tricky parts are around the public facing API: adding the warning when importing from `core` and the C-API changes in `_import_array` and friends. Is that code tested? I am worried that this change will be very noisy for downstream libraries. Do we have a handle on how many changes scipy/numexpr/numba would have to do to get clean tests?

@mattip For importing from `core` I added `test_core_shims_coherence` test, which verifies that all "semi-public" submodules in `_core` are still importable from `core` with a warning.

For `import_array`, it looks that for libs compiled with 1.x and running with nightly numpy the worst that can happen is a warning. A case where a lib was compiled with nightly and is running 1.x will work without a warning. 

For `import_array` we went through some PRs in downstream libraries to ensure building with numpy nightly will work:
- For scipy: https://github.com/scipy/scipy/pull/19275. Some time ago I tested scipy by getting a nightly and running tests against my local numpy with `_core` and warnings were the only issues (thus a warning filter was added).
- For pybind11: https://github.com/pybind/pybind11/pull/4857 so it will work in both directions (numpy version is checked and `core`/`_core` is selected appropriately).
- For numba I see that there's only [one line](https://github.com/search?q=repo%3Anumba%2Fnumba+%22numpy.core%22&type=code) in utils accessing `core`, and in numexpr [there's none](https://github.com/search?q=repo%3Apydata%2Fnumexpr%20%22numpy.core%22&type=code). To repeat, for them `import_array()` from 1.x will work as we have `numpy.core` shims.

We reviewed C changes in sub-PR https://github.com/mtsokol/numpy/pull/1 that was merged here.

mtsokol commented on numpy/_core/code_generators/generate_ufunc_api.py:
```diff
@@ -30,18 +30,24 @@
+  if (numpy == NULL && PyErr_ExceptionMatches(PyExc_ModuleNotFoundError)) {
+    PyErr_Clear();
+    numpy = PyImport_ImportModule("numpy.core._multiarray_umath");
+  }
```
I see that scipy has `import_ufunc()` internally but I will check additionally locally if it imports as expected.

ngoldbaum commented on numpy/core/multiarray.py:
```diff
@@ -1,1713 +1,9 @@
+from ._utils import _raise_warning
 
-"""
+_raise_warning("*")
```
The warning I get from this is not great:

```
DeprecationWarning: `numpy.core` has been made officially private and renamed to `numpy._core`. Accessing `_core` directly is discouraged, as members of `_core` can change without any warning at any time. You should use a public module instead that exports the attribute in question. If you still would like to access an attribute from it, please use `np._core.*`.
```

Note that the suggestion is wrong (it should point to np.core.multiarray, not np.core). But also it's not giving me the name of the symbol causing this issue, which it could.

Can we replace this with a module-level __getattr__ please? Also customize the warning for submodules of `np.core`.

mtsokol replied:
Based on our discussions, file-level warnings have been removed, so file-wise imports, such as:
```python
from numpy.core.multiarray import _reconstruct
```
raise no warning (allowing undisturbed loading of 1.x pickle files).

The warning is still present when accessing via `np.core.*`. Therefore:
```python
import numpy as np
np.core.multiarray
```
gives:
```
<ipython-input-1-725e8a67566f>:1: DeprecationWarning: `numpy.core` has been made officially private and renamed to `numpy._core`. Accessing `_core` directly is discouraged, as members of `_core` can change without any warning at any time. You should use a public module instead that exports the attribute in question. If you still would like to access an attribute from it, please use `np._core.multiarray`.
```

ngoldbaum commented on numpy/core/_multiarray_umath.py:
```diff
@@ -0,0 +1,9 @@
+from numpy._core import _multiarray_umath
+from ._utils import _raise_warning
+
+_raise_warning("*")
```
same here

mtsokol replied:
Removed - explained in the previous comment.

mtsokol commented:
I removed `NumpyUnpickler` and refined `numpy.core` stubs (`numpy._core` will be backported to 1.26 in https://github.com/numpy/numpy/pull/24906).

ngoldbaum commented on doc/release/upcoming_changes/24634.change.rst:
```diff
@@ -0,0 +1,14 @@
+if you think a mistake was made and something needs to be made public.
+
+Also, ``numpy.core`` stubs were added to provide a stable way for loading
+pickled arrays, created with NumPy 1.x.
```
I'd just delete this last sentence, users shouldn't need to care about this.

mtsokol replied:
Sure - removed!

ngoldbaum commented on numpy/_core/tests/test_regression.py:
```diff
@@ -457,24 +457,20 @@ def test_pickle_py2_bytes_encoding(self):
 
         test_data = [
             # (original, py2_pickle)
-            (np.str_('\u6f2c'),
```
All these regression tests should pass without the paths being changes, can you revert these changes?

mtsokol replied:
Looking at these binary contents, they didn't change in my opinion. 
One of your suggestions was to reformat these binary strings into a single line with `# noqa` at the end - these ones are just reformatted. Should I make them multiline?

ngoldbaum replied:
Ugh, oops, I'm sorry I misread the diff, disregard this comment!

ngoldbaum commented on numpy/_core/__init__.py:
```diff
@@ -0,0 +1,173 @@
+# We used to use `np._core._ufunc_reconstruct` to unpickle. 
+# This is unnecessary, but old pickles saved before 1.20 will be using it,
+# and there is no reason to break loading them.
+def _ufunc_reconstruct(module, name):
```
This should probably stay behind in `numpy.core` since it's explicitly for pickle backward compatibility

mtsokol replied:
Sure! Moved.

ngoldbaum commented on numpy/_core/_internal.py:
```diff
@@ -0,0 +1,948 @@
+# so don't remove the name here, or you'll
+# break backward compatibility.
+def _reconstruct(subtype, shape, dtype):
+    return ndarray.__new__(subtype, shape, dtype)
```
This also should stay behind in `np.core`, I think. I don't know if there are any tests that exercise this but it would be nice to make sure this actually still works for really old pickles?

mtsokol replied:
Do you mean pickles that would use `numpy.core._internal._reconstruct`? 
I checked it and it refers to numpy versions before 1.0, as the comment says (for 1.0 it was already there for backward compatibility). So the last version where it's relevant is `0.9.8` from May 2006, whereas Python 3.0.0 was released in December 2008 (so such pickles would only be created with Python 2 that is EOL).

I tried to build numpy `0.9.8` with Python 2.7 conda environment but it complains:
```
AssertionError: hmm, why I am not inside SVN tree???
```

I looked through the codebase from that time and I didn't find any pickle files in test directories that I could take.

My solution to test the new location of legacy `_reconstruct`: 
I copied `numpy/lib/tests/data/py2-objarr.npy` to a new `py2-np0-objarr.npy` there and changed the path inside pickle to `numpy.core._internal._reconstruct` - this way the new location is tested. Would it be sufficient?

mtsokol replied:
One correction: Actually, there is a set of test for version 0 pickles! Here:
https://github.com/numpy/numpy/blob/67539a40cb13bad56a650809bf10a49e905a250d/numpy/core/tests/test_multiarray.py#L4329

So, the change of moving `_internal._reconstruct` is tested indeed.

I think that this additional file with a test can also stay.

ngoldbaum commented on numpy/core/arrayprint.py:
For this and all of the other submodules in `np.core`, can you add module-level `__getattr__` implementations saying that accessing things via `np.core` is deprecated. Along with a nice message explaining exactly what was accessed and what the migration is.

Just as an example, right now you can do this without ever seeing a warning:

```
In [1]: import numpy.core.multiarray

In [2]: numpy.core.multiarray.ndarray
Out[2]: numpy.ndarray
```

Whereas I would expect to see a deprecation warning like 

```
"`numpy.core` has been made officially private and renamed to `numpy._core`. Accessing `_core` directly is discouraged, as members of `_core` can change without any warning at any time. You should use a public module instead that exports the attribute in question. If you still would like to access this attribute without a warning despite the lack of backward compatibility guarantees, please use `numpy._core.multiarray.ndarray`.
```

mtsokol replied:
I'm not sure about this change - when I add a custom `__getattr__` for `numpy.core.multiarray` submodule that shows a warning, then:
```python
from numpy.core.multiarray import _reconstruct
```
Also gives a warning, as `_reconstruct` is accessed with `getattr(numpy.core.multiarray, "_reconstruct")` internally.

This will cause pickles to also raise warnings that we want to avoid. I think the only way to have warnings is when user does `np.core.*` which triggers `core`'s `__getattr__`, so as it's implemented right now. Do I understand it correctly?

seberg replied:
That sounds mostly right.  When imported explicitly, a submodule import will (normally?) side-step the parents module `__getattr__`.  But that I suspect that is the only special path you can (ab)use here.
Once import of the submodule finished it has the same effect as `__getattr__()` having being called: the submodule is added to the module `__dict__()` and `__getattr__()` will never be called again (even by attribute access as is the case above).

But... if you like, I suspect you can implement a warning `__getattr__()` for all symbols that are imported from elsewhere.  Note that `multiarray._reconstruct` is special there because its `__module__` is explicitly replaced to be `np.core.multiarray`.

For some of those, they will not have a `__module__`, but in that case the submodule presumably is listed earlier in `sys.modules` so pickle will have used the submodule.
(I am not 100% sure this cannot ever be broken, but I will bet it is OK in practice.)

ngoldbaum replied:
> Do I understand it correctly?

Hmm, I thought I had a version where a module-level `__getattr__` would work a few days ago while I was messing around with his, but now I can't get that to work. I'm going to try a little more to see where I'm not understanding things.

ngoldbaum commented on numpy/core/_utils.py:
```diff
@@ -0,0 +1,15 @@
+        "members of `_core` can change without any warning at any time. "
+        "You should use a public module instead that exports the attribute "
+        "in question. If you still would like to access an attribute "
+        f"from it, please use `np._core.{attr}`.",
```
You refer to the module as `numpy` above, use either that or `np` but not both.

Also looking at the text of other warnings, I don't think it's correct to include the backticks around paths. These are going to be displayed in a terminal so the markdown formatting isn't helpful.

mtsokol replied:
I agree - I updated the warning message.

ngoldbaum reviewed (commented):
Just a few more comments to tweak some of the errors and to catch some stuff that might break existing fixes for really old pickle files.

ngoldbaum commented:
OK I think this is really ready and I'd like to merge it tomorrow. @mtsokol would you still like to do some history rewriting before I merge? I think keeping the mechanical`core` to `_core` renaming in its own commit and subsequent things we did to create the stub namespace should be in separate commits, so I'd like to avoid squashing this into one commit. This will hopefully make git archeology in the future a little bit easier.

mtsokol commented:
> OK I think this is really ready and I'd like to merge it tomorrow. @mtsokol I think you mentioned you'd like to do some history rewriting before I merge?

@ngoldbaum Sure! I squashed it into five commits: 
1. Rename `numpy.core` module to `numpy.core`
2. Adjust `numpy.core` paths in the codebase to `_core`
3. `numpy.core` shims
4. The crucial C-level change with imports
5. The rest of changes, including your imports logic

ngoldbaum commented:
Thanks for pushing this through @mtsokol!

If you're reading this because you are seeing new warnings about deprecations and are unsure what to do, read on.

The `numpy.core` namespace was always intended to contain private implementation details, but this was never clearly marked. As part of the [cleanups scheduled for Numpy 2.0](https://numpy.org/neps/nep-0052-python-api-cleanup.html), this PR renamed `np.core` to `np._core` and made accessing items in `np.core` generate deprecation warnings.

The most common usage of `numpy.core` we've found is to access functionality in NumPy that is already publicly available elsewhere. To summarize [the NumPy 2.0 migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#core-namespace), the most likely fix for the warnings you are seeings is that you will need to import from the main `numpy` namespace instead of `numpy.core`.

If it turns out that you are importing something from `numpy.core` that is not publicly available, then it is likely you are making use of private numpy internals. If you would like to continue doing that despite the lack of backward compatibility guarantees from NumPy, you can import from `numpy._core` without generating any warnings.

If you feel that you are using something that should be public but is not, please [open an issue](https://github.com/numpy/numpy/issues/new/choose) describing your need.
